state/protocol/protocol_state/epochs/statemachine.go

```diff
@@ -210,24 +210,7 @@ func (e *EpochStateMachine) Build() (*transaction.DeferredBlockPersist, error) {
 // contract, it could be that this happens too late, and we should trigger epoch fallback mode.
 // No errors are expected during normal operations.
 func (e *EpochStateMachine) EvolveState(sealedServiceEvents []flow.ServiceEvent) error {
-	parentProtocolState := e.activeStateMachine.ParentState()
-
-	// perform protocol state transition to next epoch if next epoch is committed, and we are at first block of epoch
-	// TODO: The current implementation has edge cases for future light clients and can potentially drive consensus
-	//       into an irreconcilable state (not sure). See for details https://github.com/onflow/flow-go/issues/5631
-	//       These edge cases are very unlikely, so this is an acceptable implementation in the short - mid term.
-	//       However, this code will likely need to be changed when working on EFM recovery.
-	phase := parentProtocolState.EpochPhase()
-	if phase == flow.EpochPhaseCommitted {
-		if e.activeStateMachine.View() > parentProtocolState.CurrentEpochFinalView() {
-			err := e.activeStateMachine.TransitionToNextEpoch()
-			if err != nil {
-				return fmt.Errorf("could not transition protocol state to next epoch: %w", err)
-			}
-		}
-	}
-
-	dbUpdates, err := e.applyServiceEventsFromOrderedResults(sealedServiceEvents)
+	dbUpdates, err := e.evolveActiveStateMachine(sealedServiceEvents)
 	if err != nil {
 		if protocol.IsInvalidServiceEventError(err) {
 			dbUpdates, err = e.transitionToEpochFallbackMode(sealedServiceEvents)
```

Comment:
I would suggest to inline the method `transitionToEpochFallbackMode` here I think this helps readability, because then it is immediately clear that in  both cases we are eventually calling `evolveActiveStateMachine` (removing one layer of indirection)
```suggestion
			// When the happy path state machine returns an InvalidServiceEventError, we discard its state and use the fallback state machine
			// to handle the block's epoch state evolution. The fallback state machine sets the state's EFM flag and gracefully handle all
			// service events to keep the protocol alive, no matter whether the service events are incorrect, inconsistent or unexpected.
			// Once we enter EFM, the only way to return to normal is by processing an epoch recover event by the fallback state machine.
			//    Without loss of generality, we can assume that the error above is from the happy path state machine. In case of a bug, where
			// the fallback state machine was already active above, yet it returned the `InvalidServiceEventError`, we would re-execute exactly
			// that same logic below, arrive exactly at the same conclusion (fallback state machine returned an error which it shouldn't have)
			// and crash.
			e.activeStateMachine, err = e.epochFallbackStateMachineFactory()
			if err != nil {
				return fmt.Errorf("could not create epoch fallback state machine: %w", err)
			}
			dbUpdates, err = e.evolveActiveStateMachine(sealedServiceEvents)
```
In my suggestion, I already added the documentation that I think is necessary. You should be able to just delete `transitionToEpochFallbackMode`